src/Emulator/Extensions/Mocks/HPSHostController.cs

```diff
@@ -171,6 +174,92 @@ private byte[] GetBytesFromSlave(int count, TimeInterval timeInterval)
             return result;
         }
 
+        // When we need to read for more than 255 bytes, we will need this method
+        private byte[] GetLongBytes(int count, TimeInterval timeInterval) {
+            if (count <= 255) {
+                return GetBytesFromSlave(count, timeInterval);
+            }
+            var result = new byte[count];
+            int rounds = count /255;
+            int remainder = count % 255; 
+            for (int i = 0; i <= rounds; i ++) {
+                int toRead = i == rounds ? remainder : 255;
+                Array.Copy(GetBytesFromSlave(toRead, timeInterval), 0, result, i * 255, toRead);
+            }
+            return result;
+        }
+
+        public void StartSocketServer(TimeInterval timeInterval)
+        {  
+            Console.WriteLine(currentSlave.GetType().Name);
+            string path = "/tmp/i2c.sock";
+            bool needError = false;
+            if (File.Exists(path))
+            {
+                File.Delete(path);
+            }
+            var socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
+            socket.Bind(new UnixEndPoint(path));
+            socket.Listen(0);
+            Console.WriteLine("Server started waiting for client to connect...");
+            Console.WriteLine("Starting on: "+ path);
+            currentSlave.Read(0);
+            var connection = socket.Accept();
+            Console.WriteLine("Connection established");
+            // Doubleword();
+
+            // byte[] commandBuffer = new byte[500];
+            while(SocketConnected(connection)){
+                var commandBuffer = new byte[1];
+                var lengthBytes = new byte[2];
+                int length;
+                try {
+                    connection.Receive(commandBuffer, 0, commandBuffer.Length, SocketFlags.None);
```

Comment:
It's probably important to check the return value, which indicates how many bytes were read - if there isn't enough bytes yet available, then it might not read as much as you asked for, so you probably need to read in a loop until it has read as much as you asked for. I'd probably suggest creating a helper function that you can use from all the places where you receive data.

A partial read could cause things to get out-of-sync, which could explain some of the problems you've observed.

Reply:
That is true, I'll update this